File: 3568-minimum-moves-to-clean-the-classroom/3568-minimum-moves-to-clean-the-classroom.py
class Solution:
    def minMoves(self, classroom: List[str], energy: int) -> int:
        
        # Pre - process
        m, n = len(classroom), len(classroom[0])
        def inbound(x, y):
            return 0 <= x < m and 0 <= y < n

        sx, sy = -1, -1
        litters, identifier = {}, 0
        for i in range(len(classroom)):
            for j in range(len(classroom[0])):
                if classroom[i][j] == 'L':  
                    litters[(i, j)] = identifier
                    identifier += 1
                elif classroom[i][j] == 'S':
                    sx, sy = i, j
        directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        
        # BFS part
        queue = deque([(sx, sy, energy, 0)])
        jump = 0
        best = [[[-1 for _ in range(1 << len(litters) )]  for _ in range(n)] for _ in range(m)]

        while queue:
            for _ in range(len(queue)):
                curr_x, curr_y, power, mask = queue.popleft()
            
                if mask == (1 << len(litters)) - 1:
                    return jump

                for dx, dy in directions:
                    new_x, new_y = curr_x + dx, curr_y + dy
                    if inbound(new_x, new_y) and classroom[new_x][new_y] != 'X' and power:
                        new_mask = mask | (1 << litters[(new_x, new_y)]) if (new_x, new_y) in litters else mask
                        new_power = energy if classroom[new_x][new_y] == 'R' else power - 1
                        if best[new_x][new_y][new_mask] < new_power:
                            best[new_x][new_y][new_mask] = new_power
                            queue.append([new_x, new_y, new_power, new_mask])
            
            jump += 1
        return -1

File: 3568-minimum-moves-to-clean-the-classroom/3568-minimum-moves-to-clean-the-classroom.js
/**
 * @param {string[]} classroom
 * @param {number} energy
 * @return {number}
 */
var minMoves = function(classroom, energy) {
  // pre - process
    let m = classroom.length, n = classroom[0].length
    const key = (i, j) => `${i},${j}`
    const inbound = (x, y) => 0 <= x && x < m && 0 <= y && y < n
    var identifier = 0
    var litters = new Map()
    var sx = -1, sy = -1;
    for(let i = 0; i < m; i++) {
        for(let j = 0; j < n; j++) {
            if (classroom[i][j] === 'L') {
                litters.set(key(i, j), identifier);
                identifier++;
            } else if (classroom[i][j] === 'S') {
                sx = i
                sy = j
            }
        }
    }

    // BFS
    let directions = [[0, 1], [1, 0], [-1, 0], [0, -1]]
    let best = Array.from({length : m}, (_, i) => Array.from({length : n}, (_, j) => Array.from({length : 1 << litters.size}, (_, k) => -1)))
    let curr = [[sx, sy, energy, 0]]
    let jump = 0
    let next_level = []
    while(curr.length != 0) {
        
        for(let i = 0; i < curr.length; i++) {
            let [x, y, power, mask] = curr[i] // 1, 1, 3, 2
            let new_mask = 0, new_power = 0
            
            if (mask === (1 << litters.size) - 1) {
                return jump
            }
            
            for(const [dx, dy] of directions) {
                let new_x = x + dx, new_y = y + dy
                if (inbound(new_x, new_y) && power && classroom[new_x][new_y] != 'X') {
                    new_mask = classroom[new_x][new_y] == 'L' ? mask | (1 << litters.get(key(new_x, new_y))) : mask
                    new_power = classroom[new_x][new_y] == 'R' ? energy : power - 1
                    if (best[new_x][new_y][new_mask] < new_power) {
                        best[new_x][new_y][new_mask] = new_power
                        next_level.push([new_x, new_y, new_power, new_mask])
                    }
                }
            }
        }
        curr = next_level
        next_level = []
        jump++;
    }
    return -1;  
};